### PR TITLE
add high performance training strategy

### PR DESCRIPTION
**Pull request by shippingwang**
This PR
- add high performance preprocessing in PaddleClas
- add [DALI](https://github.com/NVIDIA/DALI) in PaddleClas

It Improves almost 200%~400% training speed in no mixup series model
and 350% ~ 400% training speed in mixup series model

~~please install [paddlepaddle-gpu](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/install/Tables.html#whl-dev) compiled by GCC ≥5.4~~ 
official paddle whl package works, GCC ≥5.4 environment is still required though!

and install [dali](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/installation.html#nightly-builds) at first.

- [ ] Autoaugment
- [ ] RandAugment hasn't been supported yet.

run tools/run_dali.sh to start

- Performance
> statistics time = 10th ~ 110th batch time / 100
 on NVIDIA® Tesla® V100 
'_vd' series use mixup preprocessing

|Model | batch size | cards | using cpu preprocess time(ms/batch) | using gpu preprocess time(ms/batch)|
|- |:-: |:-: |:-: |:-: |
|ResNet50|128|1|359.3|102.3|
||256|2|445.4|195.2|
|ResNet50_vd| 128| 1| 396.8|121.4|
||256|2|737.9|208.7|

MobileNetV1 (71.30%) now
_vd series
ResNet18_vd 72.61

### CODE REVIEW

File: tools/program.py [hunk 1]
```diff
@@ -149,6 +151,7 @@ def create_loss(out,
         loss = JSDivLoss(class_dim=classes_num, epsilon=epsilon)
         return loss(out[1], out[0])
 
+    print("++++++", use_mix)
```
Comment (WuHaobo) [inline tools/program.py:154]: remove the debug comments

File: tools/program.py [hunk 2]
```diff
@@ -425,8 +437,20 @@ def run(dataloader,
...
+    for idx, batch in enumerate(dataloader):
+
+        #start_time = time.time()
         metrics = exe.run(program=program, feed=batch, fetch_list=fetch_list)
+        #end_time = time.time()
+        #statistics = end_time - start_time
+        # if idx >= 10:
+        #    sta = sta+statistics
+        # if idx == 110 and int(os.getenv("PADDLE_TRAINER_ID", 0)) == 0:
+        #    print("10-100batch speed 000", sta/100, 's/batch', 'bs', config.TRAIN.batch_size)
```
Comment (WuHaobo) [inline tools/program.py:451]: remove the debug comments

File: tools/run_dali.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+export PYTHONPATH=$PWD:$PYTHONPA
```
Comment (WuHaobo) [inline tools/run_dali.sh:3]: "$PYTHONPA"  --> $PYTHONPATH

File: configs/high_performance/dali_mix.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
+mode: 'train'
+ARCHITECTURE:
```
Comment (littletomatodonkey) [inline configs/high_performance/dali_mix.yaml:2]: 文件名改成ResNet50_vd_dali.yaml？

File: tools/dali.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+    def __len__(self):
+        return self.epoch_size("Reader")
+
+
+def build(settings, mode='train'):
+    env = os.environ
+    assert settings.get('use_gpu',
+                        True) == True, "gpu training is required for DALI"
+    #assert not settings.get('use_mix'), "mixup is not supported by DALI reader"
```
Comment (littletomatodonkey) [inline tools/dali.py:145]: 把这个注释删掉吧

File: tools/dali.py [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+        "batch size must be multiple of number of devices"
+    batch_size = bs // gpu_num
+
+    image_mean = [0.485, 0.456, 0.406]
+    image_std = [0.229, 0.224, 0.225]
+    mean = [v * 255 for v in image_mean]
+    std = [v * 255 for v in image_std]
+
+    crop = 224  # settings.crop_size
+    resize_shorter = 256  # settings.resize_short_size
```
Comment (littletomatodonkey) [inline tools/dali.py:170]: 这个从dataaug文件里面去读取预处理流程或者参数吧~

File: tools/run_dali.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+
+python3 -m paddle.distributed.launch \
+    --selected_gpus="0,1,2,3" \
+    tools/train.py \
+        -c ./configs/high_performance/dali_mix.yaml \
+	-o use_mix=True \
+        -o use_dali=True
+
+#python3 -m paddle.distributed.launch \
```
Comment (littletomatodonkey) [inline tools/run_dali.sh:14]: 这个删掉吧，或者对两端脚本分别添加注释？

File: configs/high_performance/dali_mix.yaml (hunk 4 above)
Comment (shippingwang) [reply configs/high_performance/dali_mix.yaml:2]: use_dali作为一个--args怎么样

File: tools/dali.py (hunk 5 above)
Comment (shippingwang) [reply tools/dali.py:145]: 嗯 这个文档我再改下

File: tools/dali.py (hunk 6 above)
Comment (shippingwang) [reply tools/dali.py:170]: 嗯 这个测试用的

File: tools/run_dali.sh (hunk 7 above)
Comment (shippingwang) [reply tools/run_dali.sh:14]: 嗯

**Pull request closed without merging**